exonum/src/runtime/dispatcher/schema.rs

```diff
@@ -14,194 +14,239 @@
         // Checks that instance identifier doesn't exist.
         // TODO: revise dispatcher integrity checks [ECR-3743]
-        if self.service_instance_ids().contains(&spec.id)
-            || self.pending_instance_ids().contains(&spec.id)
-        {
+        if self.instances_by_id.contains(&spec.id) {
             return Err(Error::ServiceIdExists);
         }
 
         let id = spec.id;
         let name = spec.name.clone();
-        self.pending_service_instances().put(&name, spec);
-        self.pending_instance_ids().put(&id, name);
+        self.instances.put(
+            &name,
+            InstanceState {
+                spec: spec.clone(),
+                status: InstanceStatus::Pending,
+            },
+        );
+        self.instances_by_id.put(&id, name);
+        self.pending_instances.push(spec);
         Ok(())
     }
 
-    /// Adds information about started service instance to the schema.
-    pub(super) fn add_service(&mut self, spec: InstanceSpec) {
-        debug_assert!(!self.service_instances().contains(&spec.name));
-        debug_assert!(!self.service_instance_ids().contains(&spec.id));
+    /// Adds previously started service instance.
+    pub(super) fn add_active_service(&mut self, spec: InstanceSpec) -> Result<(), Error> {
+        let artifact_id = self
+            .artifacts
+            .get(&spec.artifact.name)
+            .ok_or(Error::ArtifactNotDeployed)?
+            .spec
+            .artifact;
+
+        // Checks that runtime identifier is proper in instance.
+        if artifact_id != spec.artifact {
+            return Err(Error::IncorrectRuntime);
+        }
+        // Checks that instance name doesn't exist.
+        if self.instances.contains(&spec.name) {
+            return Err(Error::ServiceNameExists);
+        }
+        // Checks that instance identifier doesn't exist.
+        // TODO: revise dispatcher integrity checks [ECR-3743]
+        if self.instances_by_id.contains(&spec.id) {
+            return Err(Error::ServiceIdExists);
+        }
 
         let id = spec.id;
         let name = spec.name.clone();
-        self.service_instances().put(&name, spec);
-        self.service_instance_ids().put(&id, name);
+        self.instances.put(
+            &name,
+            InstanceState {
+                spec: spec.clone(),
+                status: InstanceStatus::Active,
+            },
+        );
+        self.instances_by_id.put(&id, name);
+        Ok(())
+    }
+
+    // Marks pending artifacts as deployed.
+    pub(super) fn mark_pending_artifacts_as_active(&mut self) {
+        for spec in &self.pending_artifacts {
+            self.artifacts.put(
+                &spec.artifact.name.clone(),
+                ArtifactState {
+                    spec,
+                    status: ArtifactStatus::Active,
+                },
+            );
+        }
     }
+
+    /// Marks pending instances as active.
+    pub(super) fn mark_pending_instances_as_active(&mut self) {
```

Comment:
Is that correct that after invoking this method `pending_instances` index is not empty (same for `mark_pending_artifacts_as_active`)?

Reply:
Yes.